### ISSUE TITLE
加入每日签到功能(附代码)

### ISSUE THREAD

**Issue by Jeremy111952** [Licoy/wordpress-theme-puock#503]
请将下列两段加入到主题或子主题的 functions.php
**_(A) 加入 Ajax 处理（后端）_**
add_action('wp_ajax_custom_daily_checkin', 'handle_custom_daily_checkin');
function handle_custom_daily_checkin() {
    if (!is_user_logged_in()) {
        wp_send_json_error('請先登入才能簽到');
    }

    $user_id = get_current_user_id();
    $today = date('Y-m-d');
    $signed_dates = get_user_meta($user_id, 'signed_dates', true);
    if (!is_array($signed_dates)) $signed_dates = [];

    if (in_array($today, $signed_dates)) {
        wp_send_json_error('你今天已經簽到過囉 ✅');
    }

    $signed_dates[] = $today;
    update_user_meta($user_id, 'signed_dates', $signed_dates);

    $points = (int) get_user_meta($user_id, 'user_points', true);
    $points += 1;

    $yesterday = date('Y-m-d', strtotime('-1 day'));
    $streak = (int) get_user_meta($user_id, 'checkin_streak', true);
    $streak = in_array($yesterday, $signed_dates) ? $streak + 1 : 1;

    update_user_meta($user_id, 'user_points', $points);
    update_user_meta($user_id, 'checkin_streak', $streak);

    $level = floor($points / 30) + 1;

    wp_send_json_success([
        'points' => $points,
        'streak' => $streak,
        'level' => $level,
    ]);
}

**_(B) 建立签到日历 Shortcode（前端）_**
请把这段也放进 functions.php
function render_daily_checkin_calendar() {
    ob_start();
    $user_id = get_current_user_id();
    if (!$user_id) return '<p>⚠️ 請先登入才能簽到。</p>';

    $signed_dates = get_user_meta($user_id, 'signed_dates', true);
    if (!is_array($signed_dates)) $signed_dates = [];

    $current_month = date('Y-m');
    $days_in_month = date('t');
    ?>
    <style>
    .signin-calendar {
      display: grid;
      grid-template-columns: repeat(7, 1fr);
      gap: 5px;
      max-width: 350px;
      margin-top: 20px;
    }
    .signin-calendar div {
      padding: 10px;
      border: 1px solid #ccc;
      background: #f9f9f9;
      border-radius: 6px;
      text-align: center;
    }
    .signin-calendar .signed {
      background: #4caf50;
      color: white;
    }
    </style>

    <div>
      <h3>📅 每日簽到</h3>
      <p>簽到積分：<span id="user-points">--</span></p>
      <p>連續簽到：<span id="checkin-streak">--</span> 天</p>
      <p>目前等級：<span id="user-level">--</span></p>
      <button id="checkin-btn">👉 我要簽到</button>
      <div id="checkin-status" style="margin-top: 10px;"></div>
    </div>

    <div class="signin-calendar">
      <?php for ($i = 1; $i <= $days_in_month; $i++): 
        $day = str_pad($i, 2, '0', STR_PAD_LEFT);
        $date = $current_month . '-' . $day;
        $signed = in_array($date, $signed_dates) ? 'signed' : '';
      ?>
        <div class="<?= $signed ?>"><?= $i ?></div>
      <?php endfor; ?>
    </div>

    <script>
    document.getElementById('checkin-btn').addEventListener('click', function () {
      fetch('<?php echo admin_url('admin-ajax.php'); ?>?action=custom_daily_checkin', {
        method: 'POST',
        credentials: 'same-origin'
      })
      .then(res => res.json())
      .then(data => {
        if (data.success) {
          document.getElementById('user-points').textContent = data.data.points;
          document.getElementById('checkin-streak').textContent = data.data.streak;
          document.getElementById('user-level').textContent = data.data.level;
          document.getElementById('checkin-status').innerHTML = '<span style="color:green;">✅ 簽到成功！</span>';
          location.reload();
        } else {
          let msg = typeof data.data === 'string' ? data.data : (data.data.message || '❌ 簽到失敗');
          document.getElementById('checkin-status').innerHTML = '<span style="color:red;">' + msg + '</span>';
        }
      })
      .catch(err => {
        document.getElementById('checkin-status').innerHTML = '<span style="color:red;">⚠️ 無法連線：' + err.message + '</span>';
      });
    });
    </script>
    <?php
    return ob_get_clean();
}
add_shortcode('daily_checkin', 'render_daily_checkin_calendar');

2️⃣ 使用短代码插入日历
你可以在任何地方插入以下短代码来显示签到：
[daily_checkin]
✔ 可以贴到：
页面
Elementor 区块
主题的自订 HTML 区块
侧边栏（小工具）

**_✅ 进阶功能（可加上）_**
排行榜功能程式码（加入 functions.php）
// 排行榜 Shortcode
function render_checkin_leaderboard() {
    if (!is_user_logged_in()) {
        return '<p>⚠️ 請先登入查看排行榜。</p>';
    }

    // 查詢前10名積分最高會員
    $args = [
        'meta_key' => 'user_points',
        'orderby' => 'meta_value_num',
        'order' => 'DESC',
        'number' => 10,
        'fields' => ['ID', 'display_name'],
    ];
    $user_query = new WP_User_Query($args);
    $users = $user_query->get_results();

    if (empty($users)) {
        return '<p>目前還沒有會員簽到積分。</p>';
    }

    ob_start();
    ?>
    <h3>🏆 簽到積分排行榜</h3>
    <table style="width:100%;border-collapse: collapse;">
        <thead>
            <tr>
                <th style="text-align:left;border-bottom:1px solid #ccc;padding:8px;">排名</th>
                <th style="text-align:left;border-bottom:1px solid #ccc;padding:8px;">會員名稱</th>
                <th style="text-align:right;border-bottom:1px solid #ccc;padding:8px;">積分</th>
            </tr>
        </thead>
        <tbody>
        <?php 
        $rank = 1;
        foreach ($users as $user): 
            $points = (int) get_user_meta($user->ID, 'user_points', true);
        ?>
            <tr>
                <td style="padding:8px;"><?= $rank ?></td>
                <td style="padding:8px;"><?= esc_html($user->display_name) ?></td>
                <td style="padding:8px;text-align:right;"><?= $points ?></td>
            </tr>
        <?php 
        $rank++;
        endforeach; ?>
        </tbody>
    </table>
    <?php
    return ob_get_clean();
}
add_shortcode('checkin_leaderboard', 'render_checkin_leaderboard');

使用说明
把上面程式码加到你的 functions.php（跟之前的签到功能放一起）

在你想显示排行榜的页面或文章，加入短代码：
[checkin_leaderboard]
页面就会列出积分前 10 名会员名称和积分

你可以再把每日签到区块和排行榜区块放在一起，例如：
[daily_checkin]

<hr>

[checkin_leaderboard]

**Comment by Licoy** [comment]
可以提交PR进行处理
